### ISSUE TITLE
[performance] `proc getBacktrace*(): string {.noinline.}` should reuse input buffer

### ISSUE THREAD

**Issue by timotheecour** [status-im/nim-libbacktrace#3]
`getBacktrace` can be performance critical in several cases:
* in exception heavy code, with lots of thrown/caught exceptions
* when dealing with deeply nested code (eg nim compiler itself can generate deep nesting)
* when used in combination with `nimprof` which generates a lot of stacktraces for profiling purposes

## proposal
* replace `proc getBacktrace*(): string {.noinline.}` by `proc getBacktrace*(result: var string) {.noinline.}`
and make sure the buffer passed in `getBacktrace` is allocated only once all the way down to `system/excpt.nim` (the buffer would be a `{.threadvar.}` in `system/excpt.nim`)

`result = newString(btLen)`
would become a string append instead, resulting in O(1) amortized cost instead of keep reallocating on each call to `getBacktrace`

* (a bit more complex) furthermore, `get_backtrace_c` could also reuse the same buffer to avoid allocations; this is doable but takes a bit more care (all that's needed is to expose the nim `setLen` proc to the C world via exportc, so that the C code can resize the buffer as needed

**Comment by stefantalpalaru** [comment]
I'd get some `perf` data on the actual malloc() overhead before avoiding allocations.

Also, isn't code that generates that many stack traces pathological in nature? Do we want to optimise for such a case?

**Comment by timotheecour** [comment]
> Also, isn't code that generates that many stack traces pathological in nature

IIRC that's exactly what nimprof relies on; but IMO there's a lot of room for improvement for the specific case of nimprof; eg symbolication should happen only once per unique generated stacktrace as opposed to once per stacktrace

> I'd get some perf data on the actual malloc() overhead before avoiding allocations.

for sure; this is all depending on data showing there is meaningful overhead

**Comment by arnetheduck** [comment]
the fundamental performance issue here would be doing the string formatting eagerly instead of just collecting enough data to recreate it afterwards. there should be no strings at all involved in collecting a stack trace, but then you quickly get into platform-specific territory.

introducing more mutability is not going to solve that, it will however introduce new ways for the code to break and make the code harder to reason about. a stack trace is not a string, it's a list of pointers to code.

**Comment by timotheecour** [comment]
That's this other issue: https://github.com/status-im/nim-libbacktrace/issues/4

**Comment by stefantalpalaru** [comment]
How about closing this one? Nobody needed this micro-optimisation enough to implement it, in a whole year.

**Comment by timotheecour** [comment]
that just reflects low usage IMO. Feel free to close with a "Postponed" tag since the issue is still valid. Usage will likely increase when this library is usable with minimal friction and a single flag (cf other issues), in particular if it's a proper nim flag.

**Comment by stefantalpalaru** [comment]
We can always reopen it, if any profiling data appears pointing to this as significant.
